File: Documentation/3.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Querying/HowToUseSearch.cs
using System.Collections.Generic;
using System.Linq;

using Raven.Client;
using Raven.Client.Document;

namespace Raven.Documentation.Samples.ClientApi.Session.Querying
{
	public class HowToUseSearch
	{
		private interface IFoo
		{
			/*
			#region search_1
			IRavenQueryable<T> Search<T>(
				this IQueryable<T> self,
				Expression<Func<T, object>> fieldSelector,
				string searchTerms,
				decimal boost = 1,
				SearchOptions options = SearchOptions.Guess,
				EscapeQueryOptions escapeQueryOptions = EscapeQueryOptions.EscapeAll) { ... }
			#endregion
			*/
		}

		private class User
		{
			public string Id { get; set; }

			public string Name { get; set; }

			public byte Age { get; set; }

			public ICollection<string> Hobbies { get; set; }
		}

		public HowToUseSearch()
		{
			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region search_2
					var users = session
						.Query<User>("Users/ByNameAndHobbies")
						.Search(x => x.Name, "Adam")
						.Search(x => x.Hobbies, "sport")
						.ToList();
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region search_3
					var users = session
						.Query<User>("Users/ByHobbies")
						.Search(x => x.Hobbies, "I love sport", boost: 10)
						.Search(x => x.Hobbies, "but also like reading books", boost: 5)
						.ToList();
					#endregion
				}
			}
		}
	}
}

File: Documentation/3.0/Samples/java/src/test/java/net/ravendb/clientapi/session/querying/HowToUseSearch.java
package net.ravendb.clientapi.session.querying;

import java.util.List;

import net.ravendb.client.EscapeQueryOptions;
import net.ravendb.client.IDocumentSession;
import net.ravendb.client.IDocumentStore;
import net.ravendb.client.SearchOptionsSet;
import net.ravendb.client.document.DocumentStore;
import net.ravendb.client.linq.IRavenQueryable;

import com.mysema.query.annotations.QueryEntity;
import com.mysema.query.types.Path;


public class HowToUseSearch {
  @QueryEntity
  public static class User {
    private String id;
    private String name;
    private byte age;
    private List<String> hobbies;
    public String getId() {
      return id;
    }
    public void setId(String id) {
      this.id = id;
    }
    public String getName() {
      return name;
    }
    public void setName(String name) {
      this.name = name;
    }
    public byte getAge() {
      return age;
    }
    public void setAge(byte age) {
      this.age = age;
    }
    public List<String> getHobbies() {
      return hobbies;
    }
    public void setHobbies(List<String> hobbies) {
      this.hobbies = hobbies;
    }
  }

  @SuppressWarnings("unused")
  private interface IFoo<T> {
    //region search_1
    public IRavenQueryable<T> search(Path<?> fieldSelector, String searchTerms);

    public IRavenQueryable<T> search(Path<?> fieldSelector, String searchTerms, double boost);

    public IRavenQueryable<T> search(Path<?> fieldSelector, String searchTerms, double boost, SearchOptionsSet searchOptions);

    public IRavenQueryable<T> search(Path<?> fieldSelector, String searchTerms, double boost, SearchOptionsSet searchOptions, EscapeQueryOptions escapeQueryOptions);
    //endregion
  }

  @SuppressWarnings("unused")
  public HowToUseSearch() throws Exception {
    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region search_2
        QHowToUseSearch_User u = QHowToUseSearch_User.user;
        List<User> users = session
          .query(User.class, "Users/ByNameAndHobbies")
          .search(u.name, "Adam")
          .search(u.hobbies, "sport")
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region search_3
        QHowToUseSearch_User u = QHowToUseSearch_User.user;
        List<User> users = session
          .query(User.class, "Users/ByNameAndHobbies")
          .search(u.hobbies, "I love sport", 10)
          .search(u.hobbies, "but also like reading books", 5)
          .toList();
        //endregion
      }
    }
  }
}
